db/python/tables/sequence.py

```diff
@@ -279,21 +279,44 @@ async def get_sequence_type_numbers_for_project(self, project: ProjectId):
         """
 
         _query = """
-SELECT type, COUNT(*) as n
-FROM (
-    SELECT sq.type
-    FROM sample_sequencing sq
-    INNER JOIN sample s ON s.id = sq.sample_id
-    WHERE s.project = :project
-    GROUP BY s.id, sq.type
-) as s
-GROUP BY type
+            SELECT type, COUNT(*) as n
+            FROM (
+                SELECT sq.type
+                FROM sample_sequencing sq
+                INNER JOIN sample s ON s.id = sq.sample_id
+                WHERE s.project = :project
+                GROUP BY s.id, sq.type
+            ) as s
+            GROUP BY type
         """
 
         rows = await self.connection.fetch_all(_query, {'project': project})
 
         return {r['type']: r['n'] for r in rows}
 
+    async def get_sequence_type_numbers_by_batch_for_project(self, project: ProjectId):
+        """
+        This groups by batch.
+        """
+
+        _query = """
+        SELECT batch, type, count(*) AS n
+        FROM (
+            SELECT IFNULL(JSON_EXTRACT(sq.meta, '$.batch'), 'null') as batch, sq.type
```

Comment:
Is there a reason to cast it to `"null`"`, as opposed to leaving it as None, and flow it through as the Python `None` and then JSON `null`? Or if you can't have `null` keys, casting it to something like `"no-batch"` at that point?

Reply:
SQL handles a None batch field and an omitted batch field as 2 distinct groupings and I wanted to combine them at the SQL level as it's easier to do here than in Python.

Reply:
Makes sense, can you use the SQL NULL directly instead of string?

Reply:
It's weird, the SQL NULL comes about when the field is omitted, but the Python None type is a 'null' string once in SQL, unless I made a mistake in the data_generation.py step (very possible). So I converted the SQL NULL to 'null' so it merges with the Python 'null'. I couldn't find a way to make them both turn into NULL. Thoughts?

Reply:
Resolved - merge SQL null and Python None type at SQL query, then relabel as 'no-batch' at Python level